test/hotspot/jtreg/compiler/c2/TestReachabilityFence.java

```diff
@@ -0,0 +1,270 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2;
+
+import java.lang.ref.Cleaner;
+import java.lang.ref.Reference;
+import java.util.concurrent.CountDownLatch;
+
+import jdk.internal.misc.Unsafe;
+
+
+/*
+ * @test
+ * @bug 8290892
+ * @summary Tests to ensure that reachabilityFence() correctly keeps objects from being collected prematurely.
+ * @modules java.base/jdk.internal.misc
+ * @run main/othervm -Xbatch compiler.c2.TestReachabilityFence
+ */
+public class TestReachabilityFence {
+    private static final int SIZE = 100;
+
+    static final boolean[] STATUS = new boolean[2];
+
+    interface MyBuffer {
+        byte get(int offset);
+    }
+    static class MyBufferOnHeap implements MyBuffer {
+
+        private static int current = 0;
+        private final static byte[][] payload = new byte[10][];
+
+        private final int id;
+
+        public MyBufferOnHeap() {
+            // Get a unique id, allocate memory and save the address in the payload array
+            id = current++;
+            payload[id] = new byte[SIZE];
+
+            // Initialize buffer
+            for (int i = 0; i < SIZE; ++i) {
+                put(i, (byte) 42);
+            }
+
+            // Register a cleaner to free the memory when the buffer is garbage collected
+            int lid = id; // Capture current value
+            Cleaner.create().register(this, () -> { free(lid); });
+
+            System.out.println("Created new buffer of size = " + SIZE + " with id = " + id);
+        }
+
+        private static void free(int id) {
+            System.out.println("Freeing buffer with id = " + id);
+            for (int i = 0; i < SIZE; ++i) {
+                payload[id][i] = (byte)0;
+            }
+            payload[id] = null;
+
+            synchronized (STATUS) {
+                STATUS[0] = true;
+                STATUS.notifyAll();
+            }
+        }
+
+        public void put(int offset, byte b) {
+            payload[id][offset] = b;
+        }
+
+        public byte get(int offset) {
+            try {
+                return payload[id][offset];
+            } finally {
+//                Reference.reachabilityFence(this);
```

Comment:
Drive-by comment: Is this intentionally disabled?

Reply:
Good catch. Uncommented.

(FTR I wasn't able to reproduce the problem with on-heap backing storage. Choosing between removing the test case which doesn't provoke the problem and keeping it, I chose the latter.)